flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java

```diff
@@ -17,157 +17,553 @@
+
+			if (LOG.isInfoEnabled()) {
+				LOG.info("Subtask {} will start consuming seeded shard {} from sequence number {} with ShardConsumer {}" +
+					runtimeContext.getIndexOfThisSubtask(), seededShardState.getKinesisStreamShard().toString(),
+					seededShardState.getLastProcessedSequenceNum(), seededStateIndex);
 			}
 
-			// make sure any asynchronous error is noticed
-			Throwable error = this.error.get();
-			if (error != null) {
-				throw new Exception(error.getMessage(), error);
+			shardConsumersExecutor.submit(
+				new ShardConsumer<>(
+					this,
+					seededStateIndex,
+					subscribedShardsState.get(seededStateIndex).getKinesisStreamShard(),
+					subscribedShardsState.get(seededStateIndex).getLastProcessedSequenceNum()));
+		}
+
+		// ------------------------------------------------------------------------
+
+		// finally, start the infinite shard discovery and consumer launching loop;
+		// we will escape from this loop only when shutdownFetcher() or stopWithError() is called
+
+		final long discoveryIntervalMillis = Long.valueOf(
+			configProps.getProperty(
+				KinesisConfigConstants.CONFIG_SHARD_DISCOVERY_INTERVAL_MILLIS,
+				Long.toString(KinesisConfigConstants.DEFAULT_SHARD_DISCOVERY_INTERVAL_MILLIS)));
+
+		while (running) {
+			List<KinesisStreamShard> newShardsDueToResharding = discoverNewShardsToSubscribe();
+
+			for (KinesisStreamShard shard : newShardsDueToResharding) {
+				// since there may be delay in discovering a new shard, all new shards due to
+				// resharding should be read starting from the earliest record possible
+				KinesisStreamShardState newShardState =
+					new KinesisStreamShardState(shard, SentinelSequenceNumber.SENTINEL_EARLIEST_SEQUENCE_NUM.get());
+				int newStateIndex = registerNewSubscribedShardState(newShardState);
+
+				if (LOG.isInfoEnabled()) {
+					LOG.info("Subtask {} has discovered a new shard {} due to resharding, and will start consuming " +
+							"the shard from sequence number {} with ShardConsumer {}",
+						runtimeContext.getIndexOfThisSubtask(), newShardState.getKinesisStreamShard().toString(),
+						newShardState.getLastProcessedSequenceNum(), newStateIndex);
+				}
+
+				shardConsumersExecutor.submit(
+					new ShardConsumer<>(
+						this,
+						newStateIndex,
+						newShardState.getKinesisStreamShard(),
+						newShardState.getLastProcessedSequenceNum()));
 			}
-		} finally {
-			for (ShardConsumerThread<?> consumerThread : consumerThreads) {
-				if (consumerThread.isAlive()) {
-					consumerThread.cancel();
+
+			// we also check if we are running here so that we won't start the discovery sleep
+			// interval if the running flag was set to false during the middle of the while loop
+			if (running && discoveryIntervalMillis != 0) {
+				try {
+					Thread.sleep(discoveryIntervalMillis);
+				} catch (InterruptedException iex) {
+					// the sleep may be interrupted by shutdownFetcher()
```

Comment:
I think the approach is fine. running will be set to false, then the interrupt to leave the sleep.
Flink will also start interrupting the thread if its not reacting.
Also, once the next sleep has finished it will check the running flag again, leaving the loop.